### PR TITLE
Lochi/quantization tool for trt

### PR DESCRIPTION
**Pull request by chilo-ms**
**Description**: Describe your changes.
1. Add calibration support for static quantization, currently implement MinMax algorithm to generate calibration table for Yolov3 and BERT model. This calibration table can later be used as input for TensorRT INT8 static quantization.
2. Add evaluation for model prediction. Right now it integrates COCO dataset API to evaluate and get precision for Yolov3 model.

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

**Comment by jywu-msft** [comment]
can you resolve conflict first?

File: onnxruntime/python/tools/quantization/processing.py
Comment (jywu-msft) [inline onnxruntime/python/tools/quantization/processing.py:1]: should we have a better name for this file?

File: onnxruntime/python/tools/quantization/calibrate.py
Comment (jywu-msft) [inline onnxruntime/python/tools/quantization/calibrate.py:534]: extra # ?

File: onnxruntime/python/tools/quantization/quantize.py
Comment (jywu-msft) [inline onnxruntime/python/tools/quantization/quantize.py:120]: no meaningful modifications to this file?

File: onnxruntime/python/tools/quantization/processing.py
Comment (jywu-msft) [inline onnxruntime/python/tools/quantization/processing.py:18]: from where? do we need to add any license header?

File: onnxruntime/python/tools/quantization/data_reader.py
Comment (jywu-msft) [inline onnxruntime/python/tools/quantization/data_reader.py:278]: do we need to add any proper license header?

File: onnxruntime/python/tools/quantization/calibrate.py
Comment (jywu-msft) [inline onnxruntime/python/tools/quantization/calibrate.py:244]: remove commented out lines?

File: onnxruntime/python/tools/quantization/calibrate.py
Comment (jywu-msft) [inline onnxruntime/python/tools/quantization/calibrate.py:200]: leftover debug statements?

**Comment by jywu-msft** [comment]
general question: does the calibration scripts have tests? now that we have multiple folks developing, how do we ensure that we don't break something?

File: onnxruntime/python/tools/quantization/calibrate.py
Comment (jywu-msft) [inline onnxruntime/python/tools/quantization/calibrate.py:208]: does changing keepdims impact the previous use case?

File: onnxruntime/python/tools/quantization/E2E_example_model/CalTableFlatBuffers/KeyValue.py
Comment (jywu-msft) [inline onnxruntime/python/tools/quantization/E2E_example_model/CalTableFlatBuffers/KeyValue.py:1]: why is this file under E2E_example_model ?

File: onnxruntime/python/tools/quantization/E2E_example_model/CalTableFlatBuffers/TrtTable.py
Comment (jywu-msft) [inline onnxruntime/python/tools/quantization/E2E_example_model/CalTableFlatBuffers/TrtTable.py:1]: why is this under E2E_example_model ?

File: onnxruntime/python/tools/quantization/calibrate.py
Comment (chilo-ms) [reply onnxruntime/python/tools/quantization/calibrate.py:208]: Since we want to augment all the float nodes to generate cal table. And it's possible that some of the nodes' input value are 0 which may cause error. 
In reduction_ops.cc, it mentions why:
"Can't reduce on dim with value of 0 if 'keepdims' is false. Invalid output shape would be produced. input_shape:{0}"
I will modify the script to avoid this error.

File: onnxruntime/python/tools/quantization/processing.py
Comment (chilo-ms) [reply onnxruntime/python/tools/quantization/processing.py]: Rename to preprocessing.py

File: onnxruntime/python/tools/quantization/calibrate.py
Comment (chilo-ms) [reply onnxruntime/python/tools/quantization/calibrate.py]: extra # removed

File: onnxruntime/python/tools/quantization/quantize.py
Comment (chilo-ms) [reply onnxruntime/python/tools/quantization/quantize.py]: Yes. This PR doesn't modify quantize.py

File: onnxruntime/python/tools/quantization/processing.py
Comment (chilo-ms) [reply onnxruntime/python/tools/quantization/processing.py]: Add the source. https://github.com/qqwweee/keras-yolo3/blob/master/yolo3/utils.py

File: onnxruntime/python/tools/quantization/calibrate.py
Comment (chilo-ms) [reply onnxruntime/python/tools/quantization/calibrate.py]: Removed.

File: onnxruntime/python/tools/quantization/calibrate.py
Comment (chilo-ms) [reply onnxruntime/python/tools/quantization/calibrate.py]: Removed.

File: onnxruntime/python/tools/quantization/E2E_example_model/CalTableFlatBuffers/KeyValue.py
Comment (chilo-ms) [reply onnxruntime/python/tools/quantization/E2E_example_model/CalTableFlatBuffers/KeyValue.py]: move CalTableFlatBuffers to quantization tool root folder.

File: onnxruntime/python/tools/quantization/E2E_example_model/CalTableFlatBuffers/TrtTable.py
Comment (chilo-ms) [reply onnxruntime/python/tools/quantization/E2E_example_model/CalTableFlatBuffers/TrtTable.py]: move CalTableFlatBuffers to quantization tool root folder.

File: onnxruntime/python/tools/quantization/calibrate.py
Comment (yufenglee) [inline onnxruntime/python/tools/quantization/calibrate.py:34]: Why do you need it? If you want to do batch inference, you can make the get_next return batch input.

File: onnxruntime/python/tools/quantization/calibrate.py
Comment (yufenglee) [inline onnxruntime/python/tools/quantization/calibrate.py:224]: you should be able to get it with session.get_outputs()

File: onnxruntime/python/tools/quantization/calibrate.py
Comment (yufenglee) [inline onnxruntime/python/tools/quantization/calibrate.py:232]: >get_batch [](start = 42, length = 9)

the DataReader interface doesn't have this function

File: onnxruntime/python/tools/quantization/calibrate.py
Comment (yufenglee) [inline onnxruntime/python/tools/quantization/calibrate.py:253]: you don't need this. it will be done below with code:

            pairs = [
                tuple([
                    float(min(clean_merged_dict[added_node_output_names[i]])),
                    float(max(clean_merged_dict[added_node_output_names[i + 1]]))
                ]) for i in range(0, len(added_node_output_names), 2)
            ]

File: onnxruntime/python/tools/quantization/calibrate.py
Comment (yufenglee) [inline onnxruntime/python/tools/quantization/calibrate.py:470]: all those are not common interface of DataReader

File: onnxruntime/python/tools/quantization/calibrate.py
Comment (yufenglee) [inline onnxruntime/python/tools/quantization/calibrate.py:63]: >   def write_calibration_table(self): [](start = 1, length = 37)

this function is very specific. it should not be in this function.

File: onnxruntime/python/tools/quantization/calibrate.py
Comment (chilo-ms) [reply onnxruntime/python/tools/quantization/calibrate.py:34]: The batch input is something like: {input_1: [batch_size, 3, 384, 608],  image_shape:[1,2]}
The get_next return batch input would be like: [{input_1: [1, 3, 384, 608],  image_shape:[1,2]}, {input_1: [1, 3, 384, 608],  image_shape:[1,2]}, ....]

To compose proper batch input, we still need some works. That's why I suggest to add get_batch/load_batch into DataReader. And it seems that not all the models support batch inference, therefore we need to add a function to tell whether this model supports batch inference or not.

File: onnxruntime/python/tools/quantization/calibrate.py
Comment (chilo-ms) [reply onnxruntime/python/tools/quantization/calibrate.py:224]: I think here it's to get original model's outputs not the augmented model's outputs.

File: onnxruntime/python/tools/quantization/calibrate.py
Comment (chilo-ms) [reply onnxruntime/python/tools/quantization/calibrate.py:232]: I suggest to add get_batch/load_batch to DataReader in order to support batch inference. (I've replied the reason to do so on above comments)

So, you suggest to add arguments "start" and "length" for get_batch() as you mentioned get_batch [](start = 42, length = 9) ?

File: onnxruntime/python/tools/quantization/calibrate.py
Comment (chilo-ms) [reply onnxruntime/python/tools/quantization/calibrate.py:253]: Yes, I will remove those lines.

File: onnxruntime/python/tools/quantization/calibrate.py
Comment (chilo-ms) [reply onnxruntime/python/tools/quantization/calibrate.py:470]: The reason to have these functions is that I want to make existed data_reader instance capable of setting new configuration to pre-process new input data without newing another data_reader instance.
For this scenario, the dataset has 40000 images and current T4 machine's memory can't hold all the images at once. So we have to do it with stride, as generate_calibration_table() does here. 
generate_calibration_table() has one argement, data_reader, and this function doesn't know what kind of data_reader it is.
So it can't new the data_reader, but I can set the data_reader to preprocess new stride of image data.

File: onnxruntime/python/tools/quantization/calibrate.py
Comment (chilo-ms) [reply onnxruntime/python/tools/quantization/calibrate.py:63]: https://github.com/rmccorm4/tensorrt-utils/blob/20.01/classification/imagenet/ImagenetCalibrator.py#L166

Here is an example from TensorRT utils repo and its Calibrator class includes write_calibration_table() as well.

File: onnxruntime/python/tools/quantization/calibrate.py
Comment (chilo-ms) [reply onnxruntime/python/tools/quantization/calibrate.py:253]: Also, the above code gets "only size-1 arrays can be converted to Python scalars" error when float(value) where value is [].

So, I made following change:
              pairs = []
              for i in range(0, len(added_node_output_names), 2):
                  min_value = 0
                  max_value = 0
                  min_value_array = min(clean_merged_dict[added_node_output_names[i]])
                  max_value_array = max(clean_merged_dict[added_node_output_names[i + 1]])
                  if type(min_value_array) == int or min_value_array.size > 0:
                      min_value = float(min_value_array)
                  if type(max_value_array) == int or max_value_array.size > 0:
                      max_value = float(max_value_array)

                  pairs.append(tuple([min_value, max_value]))

File: onnxruntime/python/tools/quantization/calibrate.py
Comment (chilo-ms) [reply onnxruntime/python/tools/quantization/calibrate.py]: I've removed support_batch_inference() and use getattr(data_reader, "get_batch", None) and callable(data_reader.get_batch) instead.

File: onnxruntime/python/tools/quantization/E2E_example_model/e2e_user_bert_example.py
Comment (stevenlix) [inline onnxruntime/python/tools/quantization/E2E_example_model/e2e_user_bert_example.py:19]: Is this BERT example working? Should we wait until BERT INT8 perf meets the expectation before we merge it?

File: onnxruntime/python/tools/quantization/calibrate.py
Comment (stevenlix) [inline onnxruntime/python/tools/quantization/calibrate.py:160]: Will be shape inference done first on the model in order to have all tensor names included in value_infos? If it is, we need to mentioned it in the document, otherwise user won't get completed calibration table by using original model.

File: onnxruntime/python/tools/quantization/calibrate.py
Comment (yufenglee) [reply onnxruntime/python/tools/quantization/calibrate.py:59]: emm, it is an implementation of IInt8EntropyCalibrator2, and i don't think they have same design as us.
Its functionality is to dump the parameters to a file. Why not use a util function to implement it? 

---
In reply to: [541832986](https://github.com/microsoft/onnxruntime/pull/6103#discussion_r541832986) [](ancestors = 541832986)

File: onnxruntime/python/tools/quantization/calibrate.py
Comment (yufenglee) [inline onnxruntime/python/tools/quantization/calibrate.py:166]: > input value are 0. [](start = 63, length = 19)

do you mean a 1d tensor?

File: onnxruntime/python/tools/quantization/calibrate.py
Comment (yufenglee) [reply onnxruntime/python/tools/quantization/calibrate.py:431]: This logic should be handled in the DataReader. The data reader should keep the status of reading and be implemented as something like a streaming. get_next read a batch of images, and then next call of get_next remove the last batch and load another batch. You don't need to load all the 40k images at once, right?

---
In reply to: [541832012](https://github.com/microsoft/onnxruntime/pull/6103#discussion_r541832012) [](ancestors = 541832012)

File: onnxruntime/python/tools/quantization/data_reader.py
Comment (yufenglee) [inline onnxruntime/python/tools/quantization/data_reader.py:106]: you can load batch of images here, and keep the status of start_index. that's the reason why i don't think you need get_batch.

File: onnxruntime/python/tools/quantization/calibrate.py
Comment (chilo-ms) [reply onnxruntime/python/tools/quantization/calibrate.py]: I've move this logic to DataReader. As you suggest, data reader keep the status of reading and get_next() read a batch of images until all the batches in the dataset are being read.

File: onnxruntime/python/tools/quantization/calibrate.py
Comment (chilo-ms) [reply onnxruntime/python/tools/quantization/calibrate.py]: I've moved this function to quant_utils.py

File: onnxruntime/python/tools/quantization/E2E_example_model/e2e_user_bert_example.py
Comment (chilo-ms) [reply onnxruntime/python/tools/quantization/E2E_example_model/e2e_user_bert_example.py]: BERT example is working, however I agree to not to add it until BERT INT8 perf meets the expectation. So I removed it.

File: onnxruntime/python/tools/quantization/calibrate.py
Comment (chilo-ms) [reply onnxruntime/python/tools/quantization/calibrate.py:87]: Okay, I will update to document.

File: onnxruntime/python/tools/quantization/calibrate.py
Comment (chilo-ms) [reply onnxruntime/python/tools/quantization/calibrate.py:93]: 1d tensor has 0 dimension on axis.
  The code is in reduction_ops.cc:

Comment (chilo-ms) [reply onnxruntime/python/tools/quantization/calibrate.py:93]: if (axes.size() > 0) {
      prepare_reduce_metadata.output_dims = input_dims;
      for (auto axis : axes) {
        axis = HandleNegativeAxis(axis, rank);
        ORT_ENFORCE(input_dims[axis] != 0,
                      "Can't reduce on dim with value of 0 if 'keepdims' is false. "
                     "Invalid output shape would be produced. input_shape:",
                    input_shape);
        prepare_reduce_metadata.output_dims[axis] = 1;
        reduced[axis] = true;
      }
    } else {
      // no axes provided (i.e.) default axes  => reduce on all dims
      for (auto dim : input_dims) {
        ORT_ENFORCE(keepdims || dim != 0,
                    "Can't reduce on dim with value of 0 if 'keepdims' is false. "
                    "Invalid output shape would be produced. input_shape:",
                    input_shape);
        prepare_reduce_metadata.output_dims.push_back(dim == 0 ? 0 : 1);
      }
    }

File: onnxruntime/python/tools/quantization/data_reader.py
Comment (chilo-ms) [reply onnxruntime/python/tools/quantization/data_reader.py]: Yes, I've modified the code here.

File: onnxruntime/python/tools/quantization/calibrate.py
Comment (chilo-ms) [reply onnxruntime/python/tools/quantization/calibrate.py:87]: Document updated.

**Review by stevenlix — APPROVED** [review]

**Review by jywu-msft — APPROVED** [review]

**Pull request merged**
